### PR TITLE
[l2-load-balancer] New implementation of the module 

### PR DESCRIPTION
**Pull request by Beastlex**
## Description
The new module l2-load-balancer.

## Why do we need it, and what problem does it solve?
In the current implementation, LoadBalancers based on MetalLB L2 are not fault-tolerant. We are introducing the ability to deploy LoadBalancer services across all load-balancer nodes simultaneously (multiple public IPs per LoadBalancer). This approach prevents the loss of all traffic when one of the speakers becomes unavailable.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: l2-load-balancer
type: feature
summary: The new module for redundant L2 load-balancing
impact_level: default
```

### CODE REVIEW

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	l2lbservices := make([]L2LBService, 0, 4)
+	fmt.Println("We here!")
+	fmt.Println("service snap len", len(input.Snapshots["services"]))
+	for _, serviceSnap := range input.Snapshots["services"] {
+		service, ok := serviceSnap.(ServiceInfo)
+		if !ok {
+			fmt.Println("Is not ServiceInfo")
+			continue
+		}
+		l2lb, ok := findLoadBalancerInSnapshot(service.L2LoadBalancerName, input.Snapshots["l2loadbalancers"])
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:120]: mb map?

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		if !ok {
+			// skip if we don't find L2LoadBalncer
+			fmt.Println("we don't find L2LoadBalncer")
+			continue
+		}
+
+		nodes := getNodesByNodeSelector(l2lb.Spec.NodeSelector, input.Snapshots["nodes"])
+		if len(nodes) == 0 {
+			// There no any node with selector for deploy
+			fmt.Println("There no any node with selector for deploy")
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:130]: All the messages are bad.

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/381-l2-load-balancer/crds/l2lbservices.yaml [hunk 3]
```diff
@@ -0,0 +1,537 @@
+---
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/crds/l2lbservices.yaml:1]: move to crds/internal dir.

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+func applyLoadBalancerFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var l2loadbalancer L2LoadBalancer
+
+	err := sdk.FromUnstructured(obj, &l2loadbalancer)
+	if err != nil {
+		return nil, err
+	}
+
+	return l2loadbalancer, nil
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:107]: Need an info object with name, pool and nodeSelector.

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		if len(nodes) == 0 {
+			// There no any node with selector for deploy
+			fmt.Println("There no any node with selector for deploy")
+			continue
+		}
+
+		fmt.Println("service.ExternalIPCount:", service.ExternalIPCount)
+		fmt.Println("len(nodes):", len(nodes))
+
+		for i := 1; i <= service.ExternalIPCount; i++ {
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:137]:
```suggestion
		for i := 1; i <= service.ExternalIPsCount; i++ {
```

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+			l2lbservices = append(l2lbservices, L2LBService{
+				Name:          fmt.Sprintf("%s-%d", l2lb.Name, i),
+				Namespace:     service.Namespace,
+				PreferredNode: nodes[nodeIndex].Name,
+				Ports:         service.Ports,
+				Selector:      service.Selector,
+			})
+		}
+	}
+	input.Values.Set("l2LoadBalancer.internal.l2lbservices", l2lbservices)
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:148]: sort the l2lbservices before Set.

File: ee/modules/381-l2-load-balancer/images/artifact/patches/004-l2lbservice-custom-resource.patch [hunk 7]
```diff
@@ -0,0 +1,174 @@
+From 603700d2a55e303ff3aca3487f90602f59e4b49d Mon Sep 17 00:00:00 2001
+From: Alexander Zverev <beastlex@gmail.com>
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/images/artifact/patches/004-l2lbservice-custom-resource.patch:2]:
```suggestion
From: xxx
```

File: ee/modules/381-l2-load-balancer/templates/speaker/_helper.tpl [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  {{- if index $bgpAdvertisement "communities" }}
+    {{- range $c := $bgpAdvertisement.communities }}
+      {{- if index $context.Values.metallb.bgpCommunities $c }}
+        {{- $communities = append $communities (index $context.Values.metallb.bgpCommunities $c) }}
+      {{- end }}
+    {{- end }}
+  {{- end }}
+---
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/templates/speaker/_helper.tpl:16]: Don't need one.

File: ee/modules/381-l2-load-balancer/hooks/update_node_labels.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
+/*
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/update_node_labels.go:1]: node_membership_handler

**Comment by apolovov** [comment]
Need dashboards.

**Comment by apolovov** [comment]
What if we deleted L2LoadBalancer resource with alive Service?

File: ee/modules/381-l2-load-balancer/templates/controller/rbac-for-us.yaml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      - controller
+    resources:
+      - podsecuritypolicies
+    verbs:
+      - use
+  - apiGroups:
+      - network.deckhouse.io
+    resources:
+      - l2lbservices
+      - l2lbservices/status
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/templates/controller/rbac-for-us.yaml:49]:
```suggestion
      - sdninternall2lbservices
      - sdninternall2lbservices/status
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+            Описывает ресурс L2LoadBalancer для кластера.
+          properties:
+            spec:
+              properties:
+                spec:
+                  properties:
+                    addressPool:
+                      description: |-
+                        Массив диапазонов публичных IP-адресов. Каждый диапазон может быть указан с помощью CIDR.
+                        Или прямого указания, например, `10.10.10.1-10.10.10.100`.
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml:16]:
```suggestion
                        Массив диапазонов публичных IP-адресов. Каждый диапазон может быть указан с помощью CIDR или прямого указания.
```

File: ee/modules/381-l2-load-balancer/crds/l2loadbalancers.yaml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  scope: Cluster
+  versions:
+    - name: v1alpha1
+      schema:
+        openAPIV3Schema:
+          description: L2LoadBalancer is the Schema for the l2loadbalancers API
+          properties:
+            apiVersion:
+              description: |-
+                APIVersion defines the versioned schema of this representation of an object.
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/crds/l2loadbalancers.yaml:24]: Get rid of autogenerated docs.

File: ee/modules/381-l2-load-balancer/crds/l2loadbalancers.yaml [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+              type: string
+            metadata:
+              type: object
+            spec:
+              properties:
+                addressPool:
+                  items:
+                    type: string
+                  type: array
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/crds/l2loadbalancers.yaml:44]:
```suggestion
                  type: array
                  x-examples:
                  - ["10.42.0.0/24", "10.33.5.100-10.33.5.106"]
```

File: ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+		if service.AnnotationIsMissed {
+			input.LogEntry.Warnf("Annotation %s is missed for service %s in namespace %s", keyAnnotationL2BalancerName, service.Name, service.Namespace)
+			continue
+		}
+
+		l2lb, exists := l2loadbalancers[service.L2LoadBalancerName]
+		if !exists {
+			// L2LoadBalancer is not found by name
+			input.MetricsCollector.Set("d8_l2_load_balancer_irrelevant_service", 1, map[string]string{"namespace": service.Namespace, "name": service.Name}, metrics.WithGroup("d8_l2_load_balancer"))
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/discovery_l2_load_balancers.go:149]:
```suggestion
			input.MetricsCollector.Set("d8_l2_load_balancer_orphan_service", 1, map[string]string{"namespace": service.Namespace, "name": service.Name}, metrics.WithGroup("d8_l2_load_balancer"))
```

File: ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml [hunk 15]
```diff
@@ -0,0 +1,18 @@
+- name: d8.l2-load-balancer.services
+  rules:
+    - alert: L2LoadBalancerIrrelevantServiceFound
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml:3]:
```suggestion
    - alert: L2LoadBalancerOrphanServiceFound
```

File: ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+      expr: max by (namespace, name) (d8_l2_load_balancer_irrelevant_service == 1)
+      for: 5m
+      labels:
+        severity_level: "4"
+        tier: application
+      annotations:
+        plk_markup_format: "markdown"
+        plk_protocol_version: "1"
+        plk_create_group_if_not_exists__istio_irrelevant_external_services: L2LoadBalancerIrrelevantServices,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+        plk_grouped_by__istio_irrelevant_external_services: L2LoadBalancerIrrelevantServices,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/monitoring/prometheus-rules/services.yaml:13]:
```suggestion
        plk_create_group_if_not_exists__istio_irrelevant_external_services: ClusterHasL2LoadBalancerOrphanServices,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
        plk_grouped_by__istio_irrelevant_external_services: ClusterHasL2LoadBalancerOrphanServices,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Beastlex. I've started the workflow for label 'deploy/web/stage'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Beastlex. I've started the workflow for label 'deploy/web/stage'...

File: ee/modules/381-l2-load-balancer/hooks/update_service_status.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+		patch := map[string]interface{}{
+			"status": map[string]interface{}{
+				"loadBalancer": map[string]interface{}{
+					"ingress": IPsForStatus,
+				},
+				"conditions": []metav1.Condition{
+					{
+						Status:  conditionStatus,
+						Type:    "AllPublicIPsAssigned",
+						Message: fmt.Sprintf("%d public IPs from %d were assigned", assignedIPs, totalIPs),
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/hooks/update_service_status.go:79]:
```suggestion
						Message: fmt.Sprintf("%d of %d public IPs were assigned", assignedIPs, totalIPs),
```

**Review by apolovov — APPROVED** [review]

File: ee/modules/381-l2-load-balancer/docs/README_RU.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+После чего мы можем создать стандатный ресурс _Service_ с типом `LoadBalancer` и в нем с помощью аннотаций указать:
+
+* Какой _L2LoadBalancer_ использовать. Задав этим набор нод и адресов.
+* Указать необходимое количетсво IP адресов для L2-анонсирования.
+
+<div data-presentation="../../presentations/381-l2-load-balancer/basics_l2loadbalancer_new_ru.pdf"></div>
+<!--- Source: https://docs.google.com/presentation/d/1jDqC4Bhg5NMLZWaFM32bIAzqpkUo0hOkAaRzC0yKRxE/ --->
+
+Таким образом:
+* Прикложение получит не один, а несколько "публичных" IP. Данные IP потребуется прописать в качестве A-записей для публичного домена приложения. Для последующего горизонтального масштабирования потребуется добавить дополнительные узлы-балансировщики, соответствующие _Service_ будут созданы автоматически, потребуется лишь добавить их в список A-записей прикладного домена.
```
Comment (z9r5) [inline ee/modules/381-l2-load-balancer/docs/README_RU.md:30]:
```suggestion
* Приложение получит не один, а несколько "публичных" IP. Данные IP потребуется прописать в качестве A-записей для публичного домена приложения. Для последующего горизонтального масштабирования потребуется добавить дополнительные узлы-балансировщики, соответствующие _Service_ будут созданы автоматически, потребуется лишь добавить их в список A-записей прикладного домена.
```

File: ee/modules/381-l2-load-balancer/docs/README_RU.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+<!--- Source: https://docs.google.com/presentation/d/1cs1uKeX53DB973EMtLFcc8UQ8BFCW6FY2vmEWua1tu8/ --->
+
+Данный модуль помогает обойти эти ограничения. Он предоставляет новый интерфейс _L2LoadBalancer_, который:
+
+* Позволяет с помощью `nodeSelector` связать группу нод и пул IP адресов.
+
+После чего мы можем создать стандатный ресурс _Service_ с типом `LoadBalancer` и в нем с помощью аннотаций указать:
+
+* Какой _L2LoadBalancer_ использовать. Задав этим набор нод и адресов.
+* Указать необходимое количетсво IP адресов для L2-анонсирования.
```
Comment (z9r5) [inline ee/modules/381-l2-load-balancer/docs/README_RU.md:24]:
```suggestion
* Указать необходимое количество IP-адресов для L2-анонсирования.
```

File: ee/modules/381-l2-load-balancer/docs/README_RU.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+<div data-presentation="../../presentations/381-l2-load-balancer/basics_metallb_ru.pdf"></div>
+<!--- Source: https://docs.google.com/presentation/d/1cs1uKeX53DB973EMtLFcc8UQ8BFCW6FY2vmEWua1tu8/ --->
+
+Данный модуль помогает обойти эти ограничения. Он предоставляет новый интерфейс _L2LoadBalancer_, который:
+
+* Позволяет с помощью `nodeSelector` связать группу нод и пул IP адресов.
+
+После чего мы можем создать стандатный ресурс _Service_ с типом `LoadBalancer` и в нем с помощью аннотаций указать:
+
+* Какой _L2LoadBalancer_ использовать. Задав этим набор нод и адресов.
```
Comment (z9r5) [inline ee/modules/381-l2-load-balancer/docs/README_RU.md:23]:
```suggestion
* Позволяет с помощью `nodeSelector` связать группу узлов и пул IP-адресов.

После чего мы можем создать стандартный ресурс _Service_ с типом `LoadBalancer` и в нем с помощью аннотаций указать:

* Какой _L2LoadBalancer_ использовать. Задав этим набор узлов и адресов.
```

File: ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+                        Массив диапазонов публичных IP-адресов. Каждый диапазон может быть указан с помощью CIDR или прямого указания.
+                    interfaces:
+                      description: |-
+                        Список сетевых интерфесов, с которых будет производиться анонс присвоенных IP адресов.
+                        Если поле не заполнено или задан пустой массив, то будет просиходить анонс со всех
+                        интерфейсов на ноде.
+                    nodeSelector:
+                      description: |-
+                        Позволяет ограничить список узлов, с которых будет происходить анонс.
+                        Задается в виде набора лейблов для фильтрации нод.
```
Comment (z9r5) [inline ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml:24]:
```suggestion

                        Задается в виде набора лейблов для фильтрации узлов.
```

File: ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+                spec:
+                  properties:
+                    addressPool:
+                      description: |-
+                        Массив диапазонов публичных IP-адресов. Каждый диапазон может быть указан с помощью CIDR или прямого указания.
+                    interfaces:
+                      description: |-
+                        Список сетевых интерфесов, с которых будет производиться анонс присвоенных IP адресов.
+                        Если поле не заполнено или задан пустой массив, то будет просиходить анонс со всех
+                        интерфейсов на ноде.
```
Comment (z9r5) [inline ee/modules/381-l2-load-balancer/crds/doc-ru-l2loadbalancers.yaml:20]:
```suggestion
                        Список сетевых интерфесов, с которых будет производиться анонс присвоенных IP адресов.
                        
                        Если поле не заполнено или задан пустой массив, то будет просиходить анонс со всех
                        интерфейсов узла.
```

**Review by z9r5 — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]
API - approve

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Beastlex. I've started the workflow for label 'e2e/run/static'...

**Comment by z9r5** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 9870592431 9870592431-1-con-1-27 /sys/deckhouse-oss/install:pr8465 redos@158.160.34.242

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Pull request merged**